Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1

```diff
@@ -427,60 +421,134 @@ function Unregister-ServiceFabricApplicationTypeAction
         $TimeoutSec
     )
 
-    $global:operationId = $SF_Operations.UnregisterApplicationType
+    $global:operationId = $SF_Operations.GetApplicationType
+    $getAppTypeAction = { Get-ServiceFabricApplicationTypeAction -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion }
+    $getAppTypeRetryEvaluator = {
+        param($appType)
 
-    $unregisterAction = {
-        Unregister-ServiceFabricApplicationType -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion -Force -TimeoutSec $TimeoutSec
-        if (!$?)
+        # If provisioning not started, retry register
+        if(!$appType)
         {
-            throw (Get-VstsLocString -Key SFSDK_UnableToUnregisterAppType)
+            return $true
         }
-    }
-
-    $exceptionRetryEvaluator = {
-        param($ex)
-        $appType = Get-ServiceFabricApplicationTypeAction -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion
-        # If app type already unprovisioned, don't retry
-        if (!$appType)
+        # if app type is provisioned, don't retry
+        elseif($appType -and ($appType.Status -eq [System.Fabric.Query.ApplicationTypeStatus]::Available))
         {
             return $false
         }
-
-        # if unprovisioning started, wait for it to complete
-        if (($appType.Status -eq [System.Fabric.Query.ApplicationTypeStatus]::Provisioning) -or ($appType.Status -eq [System.Fabric.Query.ApplicationTypeStatus]::Unprovisioning))
+        # if app type exist and if its status has not changed to a terminal one, do retry
+        elseif($appType -and (($appType.Status -eq [System.Fabric.Query.ApplicationTypeStatus]::Provisioning) -or ($appType.Status -eq [System.Fabric.Query.ApplicationTypeStatus]::Unprovisioning)))
         {
-            Write-Host (Get-VstsLocString -Key SFSDK_ApplicationTypeUnprovisioningStarted)
-            $appType = Wait-ServiceFabricApplicationTypeTerminalStatus -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion
+            return $true
         }
+        # if app type exist and if its status has changed to a terminal one, throw
+        elseif($appType -and (($appType.Status -ne [System.Fabric.Query.ApplicationTypeStatus]::Provisioning) -and ($appType.Status -ne [System.Fabric.Query.ApplicationTypeStatus]::Unprovisioning)))
+        {
+            throw (Get-VstsLocString -Key SFSDK_RegisterAppTypeFailedWithStatus -ArgumentList @($appType.Status, $appType.StatusDetails))
+        }
```

Comment:
What if none of these conditions are met. What is the default return value?

Reply:
If you look at semantics of last 2 ELSEIF statements than you will realize that last ELSEIF is nothing but an ELSE statement. Hence this will always work as default return.

Reply:
Have done it intentionally to enhance readability of these conditions. Also, it doesn't harm in any way.